src/software/ai/play_selection_fsm.h

```diff
@@ -63,108 +53,128 @@ struct PlaySelectionFSM
+            StopState_S + Update_E[gameStatePlaying_G && enemyHasPossession_G] /
+                              setupDefensePlay_A = DefensePlayState_S,
+            StopState_S + Update_E[gameStateSetupRestart_G] / setupSetPlay_A =
+                SetPlayState_S,
+
+            OffensePlayState_S +
+                Update_E[gameStateHalted_G] / (terminateOffensePlay_A, setupHaltPlay_A) =
+                HaltState_S,
+            OffensePlayState_S +
+                Update_E[gameStateStopped_G] / (terminateOffensePlay_A, setupStopPlay_A) =
+                StopState_S,
+            OffensePlayState_S + Update_E[gameStateSetupRestart_G] /
+                                     (terminateOffensePlay_A, setupSetPlay_A) =
+                SetPlayState_S,
+            OffensePlayState_S + Update_E[enemyHasPossession_G] /
+                                     (terminateOffensePlay_A, setupDefensePlay_A) =
+                DefensePlayState_S,
+
+            DefensePlayState_S + Update_E[gameStateHalted_G] / setupHaltPlay_A =
+                HaltState_S,
+            DefensePlayState_S + Update_E[gameStateStopped_G] / setupStopPlay_A =
+                StopState_S,
+            DefensePlayState_S + Update_E[gameStateSetupRestart_G] / setupSetPlay_A =
+                SetPlayState_S,
+            DefensePlayState_S + Update_E[!enemyHasPossession_G] / setupOffensePlay_A =
+                OffensePlayState_S,
+
+            SetPlayState_S + Update_E[gameStateHalted_G] /
+                                 (resetSetPlay_A, setupHaltPlay_A) = HaltState_S,
+            SetPlayState_S + Update_E[gameStateStopped_G] /
+                                 (resetSetPlay_A, setupStopPlay_A) = StopState_S,
+            SetPlayState_S + Update_E[gameStatePlaying_G && !enemyHasPossession_G] /
+                                 (resetSetPlay_A, setupOffensePlay_A) =
+                OffensePlayState_S,
+            SetPlayState_S + Update_E[gameStatePlaying_G && enemyHasPossession_G] /
+                                 (resetSetPlay_A, setupDefensePlay_A) =
+                DefensePlayState_S,
+            SetPlayState_S + Update_E[gameStateSetupRestart_G] / setupSetPlay_A,
 
             X + Update_E = X);
     }
 
    private:
     std::shared_ptr<Strategy> strategy_;
-    std::shared_ptr<DynamicPlay> current_dynamic_play_;
-    std::shared_ptr<OffensiveFriendlyThirdPlay> offensive_friendly_third_play_;
-    std::shared_ptr<OffensiveMiddleThirdPlay> offensive_middle_third_play_;
-    std::shared_ptr<OffensiveEnemyThirdPlay> offensive_enemy_third_play_;
+    std::shared_ptr<OffensePlay> offense_play_;
+
+    enum SetPlayType
+    {
+        NONE,
+        FRIENDLY_BALL_PLACEMENT,
+        ENEMY_BALL_PLACEMENT,
+        FRIENDLY_KICKOFF,
+        ENEMY_KICKOFF,
+        FRIENDLY_PENALTY_KICK,
+        ENEMY_PENALTY_KICK,
+        FRIENDLY_FREE_KICK,
+        ENEMY_FREE_KICK,
+    };
+
+    SetPlayType current_set_play_;
```

Comment:
can we re-use the stuff defined in `play.proto`

Reply:
Will change in #3238